### PR TITLE
[sonic-config-engine]: Improve comparison between default and supported breakout modes.

### LINKED ISSUE #7958: [DPB] sonic-cfggen does not correctly match breakout modes with different default speed
Currently sonic-cfggen does a very simple comparison between breakout modes in hwsku.json and supported modes in platform.json this means if you want to have a mode with a different default speed than is defined in platform.json

i.e. 1x10G[100G,50G] vs 1x100G[50G,10G]

Then you need an independent entry in platform.json. This is not best practice for these files as it will bloat quickly and become unmaintainable.

### Steps to Reproduce

1. Open platform.json and locate a supported breakout mode for a port (i.e. 4x25G[10G] for Ethernet0)

2. Open hwsku.json and locate the default breakout mode for the port

3. Change the platform.json breakout mode to an equivalent one with a different default (i.e. 4x10G[25G])

4. Run `sonic-cfggen --preset t1 -p -H -k [SKU YOU MODIFIED]`

### Observed Results
```
admin@l-csi-3800-01:/tmp$ sudo sonic-cfggen --preset t1 -p -H -k Mellanox-SN3800-D100C12S2 --print-data
Traceback (most recent call last):
File "/usr/local/bin/sonic-cfggen", line 436, in <module>
main()
File "/usr/local/bin/sonic-cfggen", line 316, in main
(ports, _, _) = get_port_config(hwsku, platform, args.port_config, asic_id)
File "/usr/local/lib/python3.7/dist-packages/portconfig.py", line 109, in get_port_config
return parse_platform_json_file(hwsku_json_file, port_config_file)
File "/usr/local/lib/python3.7/dist-packages/portconfig.py", line 254, in parse_platform_json_file
child_ports = get_child_ports(intf, brkout_mode, platform_json_file)
File "/usr/local/lib/python3.7/dist-packages/portconfig.py", line 204, in get_child_ports
alias_list = port_dict[INTF_KEY][interface]['breakout_modes'][breakout_mode]
KeyError: '1x10G[100G,50G,40G,25G]'
admin@l-csi-3800-01:/tmp$
```

### Expected Results

Prints a config.json where the speed of the port you configured is the default speed defined in hwsku.json

### Version Info
```
SONiC Software Version: SONiC.master.140-a4b31fcb7_Internal
Distribution: Debian 10.9
Kernel: 4.19.0-12-2-amd64
Build commit: a4b31fcb7
Build date: Mon Jun 14 15:31:01 UTC 2021
Built by: sw-r2d2-bot@r-build-sonic-ci02

Platform: x86_64-mlnx_msn4700-r0
HwSKU: ACS-MSN4700
ASIC: mellanox
ASIC Count: 1
Serial Number: MT2022X08597
Uptime: 21:11:29 up 16:58,  3 users,  load average: 0.32, 0.25, 0.26

Docker images:
REPOSITORY                                        TAG                             IMAGE ID            SIZE
harbor.mellanox.com/sonic-lc-manager/lc-manager   0.0.6                           07dab0396238        554MB
docker-syncd-mlnx                                 latest                          41d9f031f17d        670MB
docker-syncd-mlnx                                 master.140-a4b31fcb7_Internal   41d9f031f17d        670MB
docker-snmp                                       latest                          93bd11cff0c5        454MB
docker-snmp                                       master.140-a4b31fcb7_Internal   93bd11cff0c5        454MB
docker-teamd                                      latest                          39519881ce7e        423MB
docker-teamd                                      master.140-a4b31fcb7_Internal   39519881ce7e        423MB
docker-nat                                        latest                          094f94d8687c        426MB
docker-nat                                        master.140-a4b31fcb7_Internal   094f94d8687c        426MB
docker-router-advertiser                          latest                          d31730f2d8a1        412MB
docker-router-advertiser                          master.140-a4b31fcb7_Internal   d31730f2d8a1        412MB
docker-platform-monitor                           latest                          6708eb3468f2        701MB
docker-platform-monitor                           master.140-a4b31fcb7_Internal   6708eb3468f2        701MB
docker-lldp                                       latest                          e699b0e15f80        452MB
docker-lldp                                       master.140-a4b31fcb7_Internal   e699b0e15f80        452MB
docker-dhcp-relay                                 latest                          9139ad1c1bfc        419MB
docker-dhcp-relay                                 master.140-a4b31fcb7_Internal   9139ad1c1bfc        419MB
docker-database                                   latest                          7dcb114a2d7a        412MB
docker-database                                   master.140-a4b31fcb7_Internal   7dcb114a2d7a        412MB
docker-orchagent                                  latest                          9f289a7b96c3        442MB
docker-orchagent                                  master.140-a4b31fcb7_Internal   9f289a7b96c3        442MB
docker-macsec                                     latest                          0ddeec697dbc        426MB
docker-macsec                                     master.140-a4b31fcb7_Internal   0ddeec697dbc        426MB
docker-sonic-telemetry                            latest                          4e07a8ec7a64        501MB
docker-sonic-telemetry                            master.140-a4b31fcb7_Internal   4e07a8ec7a64        501MB
docker-sonic-mgmt-framework                       latest                          52b04ff3a071        570MB
docker-sonic-mgmt-framework                       master.140-a4b31fcb7_Internal   52b04ff3a071        570MB
docker-fpm-frr                                    latest                          1aa221bda051        441MB
docker-fpm-frr                                    master.140-a4b31fcb7_Internal   1aa221bda051        441MB
docker-sflow                                      latest                          1540482a009e        424MB
docker-sflow                                      master.140-a4b31fcb7_Internal   1540482a009e        424MB
docker-sonic-restapi                              latest                          6f4fdb26f843        355MB
docker-sonic-restapi                              master.140-a4b31fcb7_Internal   6f4fdb26f843        355MB
```

### PR DESCRIPTION
**Pull request by oleksandrivantsiv**
closes #7958 

#### Why I did it
The previous implementation of sonic-cfggen did a simple comparison between default breakout mode in
hwsku.json and supported modes in platform.json. To set a different default speed in hwsku.json
it was required to add one more entry to supported modes in platfrom.json file:

1x10G[100G,50G] vs 1x100G[50G,10G]

The new implementation does more intelligent parsing and analysis of supported and default modes. It
allows changing default speed without adding a new entry to platform.json.

#### How I did it
Add more intelligent parsing and analysis of supported and default modes.

#### How to verify it
Run sonic-config-engine unit tests from sonic-config-engine/tests directory:
```
sonic-config-engine/tests$ pytest ./test_cfggen_platformJson.py -vvs

test_cfggen_platformJson.py::TestCfgGenPlatformJson::test_dummy_run PASSED                                                                                                    [ 16%]
test_cfggen_platformJson.py::TestCfgGenPlatformJson::test_platform_json_all_ethernet_interfaces PASSED                                                                        [ 33%]
test_cfggen_platformJson.py::TestCfgGenPlatformJson::test_platform_json_interfaces_keys PASSED                                                                                [ 50%]
test_cfggen_platformJson.py::TestCfgGenPlatformJson::test_platform_json_no_interfaces PASSED                                                                                  [ 66%]
test_cfggen_platformJson.py::TestCfgGenPlatformJson::test_platform_json_specific_ethernet_interfaces PASSED                                                                   [ 83%]
test_cfggen_platformJson.py::TestCfgGenPlatformJson::test_print_data PASSED                                                                                                   [100%]

============================================================================= 6 passed in 2.46 seconds ==============================================================================
```
Unit tests are extended to cover changes in breakout mode parsing added in this PR

#### Which release branch to backport (provide reason below if selected)

- [ ] 201811
- [ ] 201911
- [ ] 202006
- [ ] 202012
- [ ] 202106

### CODE REVIEW

**Comment by dgsudharsan** [comment]
@praveen-li @zhenggen-xu Can you please review?

**Review by dgsudharsan — DISMISSED** [review]

**Comment by zhenggen-xu** [comment]
@samaity Please review.

**Comment by oleksandrivantsiv** [comment]
@samaity Can you please review this?

**Comment by samaity** [comment]
> @samaity Can you please review this?

checking now!

**Comment by liat-grozovik** [comment]
@samaity kindly reminder. can you please provide your inputs?

**Comment by liat-grozovik** [comment]
@oleksandrivantsiv could you please check the lgtm issue so we can now mege now that all approved the change?

**Comment by liat-grozovik** [comment]
@samaity kindly reminder, can you please review ?

File: src/sonic-config-engine/portconfig.py [hunk 1]
```diff
@@ -172,85 +172,152 @@ def parse_port_config_file(port_config_file):
...
 
-            ports[intf_name]['index'] = index.split(",")[alias_position]
+        def __ne__(self, other):
+            return not self == other
 
-            parent_intf_id += step
-            alias_position += 1
-            lanes_start += step
+        def __hash__(self):
+            return hash((self.num_ports, tuple(self.supported_speed), self.num_assigned_lanes))
```
Comment (samaity) [inline src/sonic-config-engine/portconfig.py:233]: what exactly are we checking here?

File: src/sonic-config-engine/portconfig.py [hunk 2]
```diff
@@ -172,85 +172,152 @@ def parse_port_config_file(port_config_file):
...
+        self._lanes = properties ['lanes'].split(',')
+        self._indexes = properties ['index'].split(',')
+        self._breakout_mode_entry = self._str_to_entries(bmode)
+        self._breakout_capabilities = None
+
+        # Find specified breakout mode in port breakout mode capabilities
+        for supported_mode in self._properties['breakout_modes']:
+            if self._breakout_mode_entry == self._str_to_entries(supported_mode):
+                self._breakout_capabilities = self._properties['breakout_modes'][supported_mode]
+                break
```
Comment (samaity) [inline src/sonic-config-engine/portconfig.py:247]: why did you use break here? 
what will be case when `self._breakout_mode_entry != self._str_to_entries(supported_mode)`? Should we address such cases?

File: src/sonic-config-engine/portconfig.py [hunk 3]
```diff
@@ -172,85 +172,152 @@ def parse_port_config_file(port_config_file):
...
+
+        if not self._breakout_capabilities:
+            raise RuntimeError("Unsupported breakout mode {}!".format(bmode))
+
+    def _re_group_to_entry(self, group):
+        if len(group) != 6:
```
Comment (samaity) [inline src/sonic-config-engine/portconfig.py:253]: can we make `6` as global var in the beginning of this file and use it here and also in `BRKOUT_PATTERN`?

File: src/sonic-config-engine/portconfig.py [hunk 4]
```diff
@@ -172,85 +172,152 @@ def parse_port_config_file(port_config_file):
...
+        Example of match_list for some breakout_mode using regex
+            Breakout Mode -------> Match_list
+            -----------------------------
+            2x25G(2)+1x50G(2) ---> [('2', '25G', None, '(2)', '2'), ('1', '50G', None, '(2)', '2')]
+            1x50G(2)+2x25G(2) ---> [('1', '50G', None, '(2)', '2'), ('2', '25G', None, '(2)', '2')]
+            1x100G[40G] ---------> [('1', '100G', '[40G]', None, None)]
+            2x50G ---------------> [('2', '50G', None, None, None)]
+        """
+
+        groups_list = [re.match(BRKOUT_PATTERN, i).groups() for i in bmode.split("+")]
```
Comment (samaity) [inline src/sonic-config-engine/portconfig.py:273]: if regex does not match, should we do try-catch in that case?

**Comment by samaity** [comment]
@oleksandrivantsiv  reviewed the code and asked few doubts. 

Overall, looks good to me. Kudos to the intelligence behind the smooth parsing and analysis of supported and default modes and covering more corner cases 👍 

btw a quick doubt, did you build the  whole image with this PR? As I remember, there are few build test cases in `sonic-utilities` submodule which depend on the function you are deprecating via this PR?

**Comment by oleksandrivantsiv** [comment]
@samaity thanks a lot for your feedback!

Before raising a PR I build the whole images and run a CI. The changes listed here are fully backward compatible with what was implemented before. All configuration files written before should work without any changes

File: src/sonic-config-engine/portconfig.py (hunk 2 above)
Comment (oleksandrivantsiv) [reply src/sonic-config-engine/portconfig.py:248]: Here we are trying to find whether the given breakout mode is supported. We are iterating over supported modes and checking if the given breakout mode is equal to one of them. If the breakout mode is not equal to any of the supported modes we will raise an exception with an appropriated message at line 250.

File: src/sonic-config-engine/portconfig.py (hunk 1 above)
Comment (oleksandrivantsiv) [reply src/sonic-config-engine/portconfig.py:234]: BreakoutModeEntry objects are used to compare breakout modes based on their capabilities with __eq__ and __ne__ operators. 
In Python when the class implements __eq__ operator it is recommended to implement __hash__ operator as well. __hash__ operator returns integer hash id of the object that can be used to put an object in a dictionary (as a key) or set. Here hash id is calculated based on capabilities that define breakout mode.

My initial implementation didn't include this method implementation. I added to fix LGTM alerts:

```
This pull request introduces 2 alerts when merging 04be47c into fc91dae - view on LGTM.com

new alerts:

1 for Inconsistent equality and inequality
1 for Inconsistent equality and hashing
```

File: src/sonic-config-engine/portconfig.py (hunk 3 above)
Comment (oleksandrivantsiv) [reply src/sonic-config-engine/portconfig.py]: I added a new `BRKOUT_PATTERN_GROUPS=6` global variable to replace the magic number used here.
However, I didn't change the value used in `BRKOUT_PATTERN` because it has a different meaning.
Here be indicating how many groups should have as a result of regexp matching.
In `BRKOUT_PATTERN` pattern `6` indicates up to how many digits can be in each group.
It's just a coincidence that these values are the same :)

File: src/sonic-config-engine/portconfig.py (hunk 4 above)
Comment (oleksandrivantsiv) [reply src/sonic-config-engine/portconfig.py]: To raise an exception with the meaningful error message I added try-except as suggested. Thanks!

File: src/sonic-config-engine/portconfig.py (hunk 1 above)
Comment (samaity) [reply src/sonic-config-engine/portconfig.py:234]: Thanks for the clarification, @oleksandrivantsiv.

File: src/sonic-config-engine/portconfig.py (hunk 2 above)
Comment (samaity) [reply src/sonic-config-engine/portconfig.py:248]: LGTM!

File: src/sonic-config-engine/portconfig.py (hunk 3 above)
Comment (samaity) [reply src/sonic-config-engine/portconfig.py]: Aah, makes sense! 👍

**Review by samaity — APPROVED** [review]
LGTM! 👍

File: src/sonic-config-engine/portconfig.py [hunk 5]
```diff
@@ -39,7 +39,8 @@
 INTF_KEY = "interfaces"
 OPTIONAL_HWSKU_ATTRIBUTES = ["fec", "autoneg"]
 
-BRKOUT_PATTERN = r'(\d{1,3})x(\d{1,3}G)(\[(\d{1,3}G,?)*\])?(\((\d{1,3})\))?'
+BRKOUT_PATTERN = r'(\d{1,6})x(\d{1,6}G?)(\[(\d{1,6}G?,?)*\])?(\((\d{1,6})\))?'
```
Comment (qiluo-msft) [inline src/sonic-config-engine/portconfig.py:42]: >G? [](http://example.com/codeflow?start=37&length=2)

Why allow missing `G`? What is the default unit? Is there a document for the default unit?

Comment (oleksandrivantsiv) [reply src/sonic-config-engine/portconfig.py:42]: To allow to configure speed in megabytes:
```
                "1x100000[50G,40000,25000,10G]": ["Eth36"],
                "2x50G": ["Eth36/1", "Eth36/2"],
                "4x25G[10000]": ["Eth36/1", "Eth36/2", "Eth36/3", "Eth36/4"],
                "2x25000(2)+1x50G(2)": ["Eth36/1", "Eth36/2", "Eth36/3"],
                "1x50000(2)+2x25G(2)": ["Eth36/1", "Eth36/2", "Eth36/3"]
```

**Comment by liat-grozovik** [comment]
@qiluo-msft any further comment? can we move forward and merge?

**Review by qiluo-msft — APPROVED** [review]
